sql/core/src/main/java/org/apache/spark/sql/sources/v2/DataSourceV2.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2;
+
+import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap;
+import org.apache.spark.sql.sources.v2.reader.DataSourceV2Reader;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * The main interface and minimal requirement for data source v2 implementations. Users can mix in
+ * more interfaces to implement more functions other than just scan.
+ */
+public interface DataSourceV2 {
+
+  /**
+   * The main entrance for read interface.
+   *
+   * @param schema the full schema of this data source reader. Full schema usually maps to the
```

Comment:
I think the schema should be handled a little differently. We should assume that a data source has a schema that applies to all of its data, instead of passing one in. That's the case with most datasets, like Hive tables, relational tables, and Cassandra tables. I'd also argue that it is a best practice so we don't have expensive operations to infer the schema, like we do for non-Metastore tables.

So instead of passing the full schema here and having an interface to infer a schema, the data source should be expected to report its schema while analyzing the query (which may require inference). Then the schema passed to create a reader should be the expected schema, or projection schema after the query is optimized.

Also, I'd like to understand how column IDs should be handled. I'm building a data source that uses column IDs to implement schema evolution that supports, add, delete, and rename. The column IDs for a table are only unique within that table (another table can reuse them), so it doesn't necessarily make sense to expose them to Spark. Spark will still need its own unique attribute IDs. But, I'd prefer to have Spark request columns by ID rather than by name so that Spark handles name resolution between a query and the data source.

I think what makes sense is for the data source to create a Spark schema with attributes that Spark will use in the projection schema. So if my data source has columns `0:x` and `1:y`, I create a Spark schema with those columns and they are assigned Spark attribute IDs in the process, say `x#67` and `y#68`. When Spark requests a reader, the projection schema uses those attributes, which I can then map back to internal IDs. Spark should call `createReader` with a schema containing `x#67`, not `x#104`. Sound reasonable?

Reply:
> So instead of passing the full schema here and having an interface to infer a schema, the data source should be expected to report its schema while analyzing the query (which may require inference).

I'm +1 on it if we can rewrite the data source part from scratch...
But now, since users can specify a schema in `DataFrameReader`, which means the data source should be able to take a user specified schema.

For the column id stuff, I don't totally understand it. Attribute id is internal to Spark, Spark only uses column names when interacting with data sources, that's why we only use `StructType` in various APIs.

Reply:
@rdblue 

"We should assume that a data source has a schema that applies to all of its data, instead of passing one in. That's the case with most datasets, like Hive tables, relational tables, and Cassandra tables."

This is not true though. Hive table, for example, the schema is specified in the catalog, which means it has to be passed into the underlying file-based data source. The file-based data source itself is actually not dictating the schema.

Reply:
> This is not true though. Hive table, for example, the schema is specified in the catalog, which means it has to be passed into the underlying file-based data source.

I think what I didn't understand is that the datasource API is between the metastore and the files in a table. So the metastore is queried to get a schema, then the data files are treated as schema-on-read. But why should that be the case? I would rather be able to have Hive tables entirely behind the DataSource API, so they can be handled the same way as a JDBC table or another source with a pre-determined schema for its data.

With that missing context, I can see how the interface proposed here looks designed for schema-on-read, rather than for tables that already have a fixed schema: it assumes that the data source will commonly use an externally provided schema, can't necessarily infer one (e.g. JSON), and the projection is separate from the data source schema. I think it's reasonable to consider whether this API should be based on schema-on-read or should assume that the data source has a schema along the lines of what I was suggesting before.

The proposal doc lists 3 cases: an external schema is required, schema is fixed and a user schema is not allowed, and a user schema is optional because it can be inferred.

For sources with fixed schemas, `DataSourceV2SchemaProvider#inferSchema` is required, but that also pulls in `acceptsUserDefinedSchema` that defaults to true. A fixed-schema source also has to implement `createReader` with a "full schema of the data source" passed in. Those two imply that fixed-schema sources should support schema-on-read, but I think that's a bad idea. The problem is similar to why we don't want to pass a map of options: it makes the source implementation decide how to reconcile the user-defined schema with the source schema, and that probably won't be consistent across sources. I think that's a good reason to separate out the methods needed for schema-on-read to a different interface.

What about a `SchemaOnRead` interface that can be used to set the schema instead of passing it to the `createReader` method? Then `acceptsUserDefinedSchema` can be dropped from the provider interface because it is signalled by `SchemaOnRead`. (I'd also rename `inferSchema` to `getSchema`.) Then, `createReader` here would drop the schema argument. We could also have a `createReader` with a schema argument that passes the requested projection; that's where I would expect to pass the requested schema.

For the other two cases, the proposed API works okay. It's a little strange that those that require a user schema won't implement `DataSourceV2SchemaProvider`, which is what signals that a user-defined schema is accepted even though one can be passed to any data source. A `SchemaOnRead` interface would be slightly better for the required case as well by cleaning this up.